subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/execution/Combinators.kt

```diff
@@ -18,21 +18,69 @@ package org.gradle.kotlin.dsl.execution
 
 import org.jetbrains.kotlin.lexer.KotlinLexer
 import org.jetbrains.kotlin.lexer.KtToken
+import org.jetbrains.kotlin.lexer.KtTokens
 import org.jetbrains.kotlin.lexer.KtTokens.CLOSING_QUOTE
 import org.jetbrains.kotlin.lexer.KtTokens.COMMENTS
 import org.jetbrains.kotlin.lexer.KtTokens.IDENTIFIER
+import org.jetbrains.kotlin.lexer.KtTokens.LBRACE
+import org.jetbrains.kotlin.lexer.KtTokens.LBRACKET
 import org.jetbrains.kotlin.lexer.KtTokens.LPAR
 import org.jetbrains.kotlin.lexer.KtTokens.OPEN_QUOTE
+import org.jetbrains.kotlin.lexer.KtTokens.RBRACE
+import org.jetbrains.kotlin.lexer.KtTokens.RBRACKET
 import org.jetbrains.kotlin.lexer.KtTokens.REGULAR_STRING_PART
 import org.jetbrains.kotlin.lexer.KtTokens.RPAR
 import org.jetbrains.kotlin.lexer.KtTokens.SEMICOLON
 import org.jetbrains.kotlin.lexer.KtTokens.WHITE_SPACE
+import kotlin.reflect.KProperty
 
 
 internal
 typealias Parser<T> = KotlinLexer.() -> ParserResult<T>
 
 
+@Suppress("unused")
+private
+val debugger: ParserDebugger = ParserDebugger()
+
+
+@Suppress("unused")
+private
+class ParserDebugger {
+    var level = 0
+
+    fun <T> debug(name: String, parser: Parser<T>): Parser<T> = {
+        val levelString = "\t".repeat(level)
+        println("${levelString}Parsing with $name @ ${this.currentPosition.offset} ...")
+        level++
+        val result = parser()
+        level--
+        println("${levelString}Parsing with $name done @ ${this.currentPosition.offset}, ${if (result is ParserResult.Success) "successful (${result.result})" else "failed"}")
+        result
+    }
+}
+
+
+internal
+class DebugDelegate<T>(val parserBuilder: () -> Parser<T>) {
+
+    operator fun getValue(thisRef: Any?, property: KProperty<*>): Parser<T> =
+//        debugger.debug(property.name, parserBuilder())
+        parserBuilder()
+}
+
+
+internal
+fun <T> debug(parserBuilder: () -> Parser<T>) = DebugDelegate(parserBuilder)
+
+
+internal
+fun <T> debugReference(parserBuilder: () -> Parser<T>): Parser<T> {
```

Comment:
🗒️ This deserves some documentation. When would I want to use this?

Reply:
Yes, it's a weird thing and not necessarily of generic usefulness. I'm moving it closer to its usage.